prov/sm2/src/sm2_ep.c

```diff
@@ -91,379 +89,215 @@ int sm2_getname(fid_t fid, void *addr, size_t *addrlen)
 	*addrlen = strlen(ep->name) + 1;
 
 	if (!ret)
-		((char *) addr)[*addrlen - 1] = '\0';
+		((char *)addr)[*addrlen - 1] = '\0';
+
+	*addrlen = SM2_NAME_MAX;
```

Comment:
This is worthy of attention.  This fixed a startup race condition in OMPI + SM2.  OMPI takes the addrlen returned by fi_getname and creates a char[num_ranks][addrlen], and memcpys in addrlen into the char* (assuming the string fits, and is null terminated).  SM2 coordination was using strlen() to determine where one string ended, and the next string began.  By setting the size to a constant, both work well.  This approach seems to also align with the docs (but there might also be a bug in OMPI):


fi_getname / fi_getpeer
The fi_getname and fi_getpeer calls may be used to retrieve the local or peer endpoint address, respectively. On input, the addrlen parameter should indicate the size of the addr buffer. If the actual address is larger than what can fit into the buffer, it will be truncated and -FI_ETOOSMALL will be returned. **On output, addrlen is set to the size of the buffer needed to store the address, which may be larger than the input value.**

fi_getname is not guaranteed to return a valid source address until after the specified endpoint has been enabled or has had an address assigned. An endpoint may be enabled explicitly through fi_enable, or implicitly, such as through fi_connect or fi_listen. An address may be assigned using fi_setname. fi_getpeer is not guaranteed to return a valid peer address until an endpoint has been completely connected – an FI_CONNECTED event has been generated.

Reply:
I agree with Luke. This seems like an Open MPI issue. We found the same issue in OneCCL and had to fix it in there.

Reply:
@lrbison is going to own the OMPI bugfix